device/MatrixMultiplication.cpp

```diff
@@ -187,48 +326,44 @@ void Compute(hlslib::Stream<PackedFloat> &a_in, hlslib::Stream<PackedFloat> &b_i
     PackedFloat a_buffer;  // Just to make A symmetric to B and C
     PackedFloat b_buffer[kTileSizeM];
     PackedFloat c_buffer[kTileSizeN * kTileSizeM];
-    for (int n0 = 0; n0 < hlslib::CeilDivide(size_n, kTileSizeN); ++n0) {
-        for (int m0 = 0; m0 < hlslib::CeilDivide(size_m, kTileSizeM); ++m0) {
+    const int tiles_n = hlslib::CeilDivide(size_n, kTileSizeN);
+    const int tiles_m = hlslib::CeilDivide(size_m, kTileSizeM);
+Compute_TilesN:
+    for (int n0 = 0; n0 < tiles_n; ++n0) {
+    Compute_TilesM:
+        for (int m0 = 0; m0 < tiles_m; ++m0) {
+        Compute_K:
             for (int k = 0; k < size_k; ++k) {
+            Compute_N:
                 for (int n1 = 0; n1 < kTileSizeN; ++n1) {
+                Compute_M:
                     for (int m1 = 0; m1 < kTileSizeM; ++m1) {
 #pragma HLS PIPELINE II = 1
 #pragma HLS LOOP_FLATTEN
-                        PackedFloat a, b, c;
-                        if (m1 == 0) {
-                            a = a_in.Pop();
-                            a_buffer = a;
-                        } else {
-                            a = a_buffer;
-                        }
-                        if (n1 == 0) {
-                            b = b_in.Pop();
-                            b_buffer[m1] = b;
-                        } else {
-                            b = b_buffer[m1];
-                        }
-                        if (k == 0) {
-                            c = c_in.Pop();
-                        } else {
-                            c = c_buffer[n1 * kTileSizeM + m1];
-                        }
+                        const PackedFloat a_read = a_in.Pop();
+                        const PackedFloat b_read = b_in.Pop();
+                        const PackedFloat c_read = c_in.Pop();
+                        const PackedFloat a = (m1 == 0) ? a_read : a_buffer;
+                        const PackedFloat b = (n1 == 0) ? b_read : b_buffer[m1];
+                        const PackedFloat c = (k == 0) ? c_read : c_buffer[n1 * kTileSizeM + m1];
+                        a_buffer = a;
+                        b_buffer[m1] = b;
                         // Ignore contributions from out-of-bound indices
                         const bool in_bounds = (n0 * kTileSizeN + n1 < size_n) && (m0 * kTileSizeM + m1 < size_m);
                         // Meat of the computation
                         const auto res = in_bounds ? MultiplyAccumulate(a, b, c) : c;
```

Comment:
Somewhat of a micro-optimization but it would probably be better to pass 0 into the MAC instead of bypassing the MAC completely. So like
MultiplyAccumulate(in_bounds ? a : 0, in_bounds ? b : 0, c) : c;
That way we don't need a bunch of registers that have to bypass the multiplier